bitwarden_license/src/Commercial.Infrastructure.EntityFramework/SecretsManager/Repositories/SecretRepository.cs

```diff
@@ -293,6 +293,21 @@ public async Task<(bool Read, bool Write)> AccessToSecretAsync(Guid id, Guid use
         return (policy.Read, policy.Write);
     }
 
+    public async Task EmptyTrash(DateTime currentDate, uint deleteAfterThisNumberOfDays)
+    {
+        using (var scope = ServiceScopeFactory.CreateScope())
+        {
+            var dbContext = GetDatabaseContext(scope);
+            var secrets = dbContext.Secret.Where(s => s.DeletedDate != null && s.DeletedDate < currentDate.AddDays(-deleteAfterThisNumberOfDays));
+            await secrets.ForEachAsync(secret =>
+            {
+                dbContext.Attach(secret);
+                dbContext.Remove(secret);
+            });
+            await dbContext.SaveChangesAsync();
```

Comment:
We should verify that this is a `DELETE FROM  ...` query without any loading.

Reply:
Are you recommending something like

```csharp
var secrets = dbContext.Secret.Where(s => s.DeletedDate != null && s.DeletedDate < currentDate.AddDays(-deleteAfterThisNumberOfDays));

dbContext.RemoveRange(secrets);
```

To prevent a select and then loop, generated by EF Core?

Reply:
@Hinton what's the goal of ensuring there is no loading? More performance if there are many secrets to delete?

I have captured logging when the job runs using the existing `EmptyTrash` implementation in the PR, as well as using the following implementation:

```csharp
using var scope = ServiceScopeFactory.CreateScope();
var dbContext = GetDatabaseContext(scope);

var secrets = dbContext.Secret.Where(s => s.DeletedDate != null && s.DeletedDate < currentDate.AddDays(-deleteAfterThisNumberOfDays));

dbContext.Secret.RemoveRange(secrets);

await dbContext.SaveChangesAsync();
```

In both cases, the captured logs seem to indicate the entities are being loaded: `Context 'DatabaseContext' started tracking 'Secret' entity.` I believe this is because the queries are [tracked by default](https://learn.microsoft.com/en-us/ef/core/what-is-new/ef-core-7.0/whatsnew#basic-executedelete-examples).

Is there another way to delete entities without loading? [No-tracking](https://learn.microsoft.com/en-us/ef/core/querying/tracking#tracking-queries) seems to only be useful during read-only queries. [ExecuteDelete](https://learn.microsoft.com/en-us/ef/core/what-is-new/ef-core-7.0/whatsnew#basic-executedelete-examples) seems promising but is available starting in EF Core 7.0, which we would have to upgrade for.

We could use `ExecuteSqlRaw()` or a stored procedure? Let me know if you have a better way!

Reply:
We might want to look into upgrading EF to 7 (supports .net 6) since it supports this, https://learn.microsoft.com/en-us/ef/core/what-is-new/ef-core-7.0/whatsnew#executeupdate-and-executedelete-bulk-updates.

If that's out of scope or we want it before that is available, A stored procedure is probably the way to go. RemoveRange will load everything to memory (although only need 2 queries to my understanding, the select and delete).

Reply:
We are now using EF Core 7 so we can use ExecuteDelete functions 🎉

https://github.com/bitwarden/server/pull/3006